### PR TITLE
Updated dependencies, new audio resampling library, performance improvements and volume control

### LINKED ISSUE #9: Volume Control not working
Hi & Merry Christmas! This Bot is exactly what I was looking for as a middleware for a bigger project, but I ran into a 
tiny tiny problem.... when trying to adjust the volume through spotify it doesn't seem to work. 

Unfortunately, I'm not into rust but still went ahead and tried to rebuild the bot with initial volume 
for librespot and also adjusting the volume in the enable_connect function, however, it doesn't affect the bot as well.
Volume remains constant. 

Is there a way to adjust the initial volume of the bot?

### PR DESCRIPTION
**Pull request by sibartel**
Heyho @codetheweb,

as promised some days ago in issue #9, here are my changes open to be discussed:
- Updated dependencies:
  - [librespot](https://github.com/librespot-org/librespot) from [0.2.0](https://github.com/librespot-org/librespot/releases/tag/v0.2.0) to [0.3.1](https://github.com/librespot-org/librespot/releases/tag/v0.3.1)
  - [songbird](https://github.com/serenity-rs/songbird) from [0.1.6](https://github.com/serenity-rs/songbird/releases/tag/v0.1.6) to [0.2.0](https://github.com/serenity-rs/songbird/releases/tag/v0.2.0)
- Switched the library for audio resampling from [samplerate](https://github.com/Prior99/rust-samplerate) to [rubato](https://github.com/HEnquist/rubato):

  [Samplerate](https://github.com/Prior99/rust-samplerate) is based on a c library and the function call used is not recommended for streamed audio because it does not work over chunk boundaries. As written in the [docs](https://docs.rs/samplerate/latest/samplerate/fn.convert.html):

  > Perform a simple samplerate conversion of a large chunk of audio. This calls src_simple of libsamplerate which is not suitable for streamed audio.

  Therefore, this switch should make aoede more portable and additionally improve the audio quality.

- Performance improvements:
  - Implemented non blocking EmittedSink::read.
  - Increased EmittedSink internal sync_channel buffer size to fit at least one resampling frame. And therefore, most times no synchronisation is necessary.
  - Changed datatype in internal sync_channel to fit the floats of both channels. This reduces the amount of synchronisations needed between EmittedSink::read and EmittedSink::write.
- Implemented volume control using librespot::SoftMixer. Should close #9.

I'm looking forward to hear from you.
Best regards!

### CODE REVIEW

File: src/lib/player.rs [hunk 1]
```diff
@@ -35,71 +39,72 @@ pub struct SpotifyPlayer {
...
+    pub receiver: Arc<Mutex<Receiver<[f32; 2]>>>,
+    input_buffer: Arc<Mutex<(Vec<f32>, Vec<f32>)>>,
+    resampler: Arc<Mutex<FftFixedInOut<f32>>>,
+    resampler_input_frames_needed: usize,
 }
 
 impl EmittedSink {
     fn new() -> EmittedSink {
-        let (sender, receiver) = sync_channel::<u8>(64);
+        let (sender, receiver) = sync_channel::<[f32; 2]>((1024f32 * 1.2).ceil() as usize);
```
Comment (codetheweb) [inline src/lib/player.rs:51]: Is this saying that the upper bound is 1.2 * the amount of data in a single frame? If so, could you just add a comment explaining it?

File: src/lib/player.rs [hunk 2]
```diff
@@ -108,15 +113,49 @@ impl audio_backend::Sink for EmittedSink {
...
         let receiver = self.receiver.lock().unwrap();
 
-        #[allow(clippy::needless_range_loop)]
-        for i in 0..buff.len() {
-            buff[i] = receiver.recv().unwrap();
+        let chunk_size = mem::size_of::<f32>() * 2;
+        let mut bytes_written = 0;
+        while bytes_written + (chunk_size - 1) < buff.len() {
+            if bytes_written == 0 {
+                // We can not return 0 bytes because songbird then things that the track has ended,
```
Comment (codetheweb) [inline src/lib/player.rs:123]:
```suggestion
                // We can not return 0 bytes because songbird then thinks that the track has ended,
```

**Review by codetheweb — CHANGES_REQUESTED** [review]
Thank you, this approach makes much more sense than what I was doing. Downsides of learning a language by writing something in it. 😛 

Looks like CPU usage while streaming dropped from 10-12% to 5-6% which is excellent as well. 👍

File: src/lib/player.rs (hunk 1 above)
Comment (sibartel) [reply src/lib/player.rs]: Yes. I added a comment explaining the meaning and replaced the vague calculation with the hardcoded value for our used chunk size and frequencies.

**Comment by sibartel** [comment]
Heyho @codetheweb,
happy new year! Thx for the review, I made the following changes:
- Added error to inform about minimum buffer size for EmittedSink::read call
- Fixed spelling
- Added comment to explain the upper bound of sync_channel
- Replaced vague calculation of sync_channel upper bound with the correct value for our settings

Should i squash the commits in some ways?

Regarding the CPU usage: Using the release build I have permanently less than 0.3% on my machine. Maybe we can investigate and optimize the cpu usage further in the future.

**Comment by codetheweb** [comment]
Looks good!

Happy New Year to you too.

I'll squash when merging.

> Regarding the CPU usage: Using the release build I have permanently less than 0.3% on my machine. Maybe we can investigate and optimize the cpu usage further in the future.

I should have clarified that my usage stats were assuming full utilization of one core is 100%. If you're measuring based on overall utilization of the CPU, that would explain the difference. I'm on a M1 Max. I'm happy with performance as-is, but if you think it can further be optimized go for it. :)

**Review by codetheweb — APPROVED** [review]

**Comment by codetheweb** [comment]
Released in v0.3.0.

**Pull request merged**